governance/program/src/addins/voter_weight.rs

```diff
@@ -49,6 +68,17 @@ pub struct VoterWeightRecord {
     /// As a common pattern Revise instruction to update the weight should be invoked before governance instruction within the same transaction
     /// and the expiry set to the current slot to provide up to date weight
     pub voter_weight_expiry: Option<Slot>,
+
+    /// The governance action the voter's weight pertains to
+    /// It allows to provided voter's weight specific to the particular action the weight is evaluated for
+    /// When the action is provided then the governance program asserts the executing action is the same as specified by the addin
+    pub weight_action: Option<VoterWeightAction>,
+
+    /// The target the voter's weight  action pertains to
+    /// It allows to provided voter's weight specific to the target the weight is evaluated for
+    /// For example when addin supplies weight to vote on a particular proposal then it must specify the proposal as the action target
+    /// When the target is provided then the governance program asserts the target is the same as specified by the addin
+    pub weight_action_target: Option<Pubkey>,
```

Comment:
So are these fields just a safeguard to be sure that the addin program behaves correctly and is forced to acknowledge the parameters passed in?  I went back and forth on this, but finally it seems like a good idea.  If they're set in correctly, it means the addin program actually deserialized what you passed in, which is better than nothing.

Reply:
It's a reversed relationship. This account is owned by the `addin` and it's the `addin` responsibility to set these fields and then the account is passed to `spl-governance` which validates they match the invoked instruction.

This way the `addin` can for example fetch and inspect `Proposal` and specify in `VoterWeightRecord`  the `voter_weight` was provided for voting on that particular proposal. Then `spl-governance`  checks that in fact `CastVote` was invoked for the `Proposal`

If the `addin` doesn't specify the acton and the target then it means the weight is not context sensitive and can be used for any action and any target